In [1]:
import pandas as pd

In [2]:
%load_ext sql

In [3]:
%sql mysql://USERNAME:PASSWORD@HOST/DATABASE

## Exploratory Data Analysis

1. Determine customer count and total amount spent by state to see where most sales are being made.

In [18]:
%%sql
SELECT
	state,
	COUNT(cust_id) AS num_customers,
	ROUND(SUM(order_total), 2) AS total_spent,
	SUM(order_quantity) AS total_units_purchased
FROM customer_orders co 
JOIN customers c ON co.cust_id = c.id
GROUP BY state
ORDER BY total_spent DESC
LIMIT 10;

 * mysql://admin:***@lmu-dev-sql.c7huj6y1fkyn.us-east-2.rds.amazonaws.com/sql_portfolio_project
10 rows affected.


[('TX', 2288, Decimal('147636.36'), Decimal('4764')),
 ('CA', 2175, Decimal('140043.81'), Decimal('4519')),
 ('FL', 1340, Decimal('84943.59'), Decimal('2741')),
 ('NY', 935, Decimal('61143.27'), Decimal('1973')),
 ('GA', 789, Decimal('53333.79'), Decimal('1721')),
 ('PA', 772, Decimal('49924.89'), Decimal('1611')),
 ('VA', 744, Decimal('46732.92'), Decimal('1508')),
 ('AL', 676, Decimal('45865.20'), Decimal('1480')),
 ('NC', 619, Decimal('41309.67'), Decimal('1333')),
 ('IL', 541, Decimal('35390.58'), Decimal('1142'))]

Analyis: California, Texas, and Florida are the states with the most purchases, so we shou

2. Find the top categories and the sales details of each category.

In [19]:
%%sql
SELECT
	p.category,
	ROUND(SUM(order_total), 2) AS total_spent,
	SUM(order_quantity) AS units_ordered,
	SUM(p.review_count) AS total_reviews
FROM product p 
JOIN customer_orders co
	ON p.asin = co.asin
GROUP BY p.category
ORDER BY total_spent DESC;

 * mysql://admin:***@lmu-dev-sql.c7huj6y1fkyn.us-east-2.rds.amazonaws.com/sql_portfolio_project
9 rows affected.


[('Toys & Games', Decimal('202302.72'), Decimal('6528'), Decimal('126268339')),
 ('Pet Supplies', Decimal('200815.20'), Decimal('6480'), Decimal('113061162')),
 ('Appliances', Decimal('163100.37'), Decimal('5263'), Decimal('95406835')),
 ('Sports & Outdoors', Decimal('162077.70'), Decimal('5230'), Decimal('140186167')),
 ('Beauty & Personal Care', Decimal('142089.15'), Decimal('4585'), Decimal('157401315')),
 ('Clothing, Shoes & Jewelry', Decimal('127523.85'), Decimal('4115'), Decimal('69731355')),
 ('Electronics', Decimal('114291.12'), Decimal('3688'), Decimal('35048733')),
 ('Office Products', Decimal('97463.55'), Decimal('3145'), Decimal('51670707')),
 ('Home & Kitchen', Decimal('88848.33'), Decimal('2867'), Decimal('26738254'))]

Analysis: Upon further analysis and due to the method by which customer and order data was generated, I cannot gain much information by looking into the "category" column. As I was unable to get the actual categories for the products (Amazon blocks webscraping attempts and I would have to upgrade to the "premium" tier to make all the API requests), I assigned each product a category randomly.

3. Find out which products are ordered the most, as well as compare it to the number of reviews.

In [30]:
%%sql
SELECT
	p.asin,
	p.title,
    p.current_price,
	p.category,
	COUNT(co.order_id) AS orders,
	p.review_count
FROM product p 
JOIN customer_orders co 
	ON p.asin = co.asin
GROUP BY p.asin
ORDER BY orders DESC
LIMIT 20;

 * mysql://admin:***@lmu-dev-sql.c7huj6y1fkyn.us-east-2.rds.amazonaws.com/sql_portfolio_project
20 rows affected.


[('B07XJ8C8F7', 'Echo Dot (4th Gen) | Smart speaker with clock and Alexa | Glacier White', Decimal('59.99'), 'Beauty & Personal Care', 682, 175857),
 ('B0168CNYRE', 'Dash Rapid Egg Cooker: 6 Egg Capacity Electric Egg Cooker for Hard Boiled Eggs, Poached Eggs, Scrambled Eggs, or Omelets with Auto Shut Off Feature - Red', Decimal('17.99'), 'Appliances', 415, 97066),
 ('B07VGRJDFY', 'Nintendo Switch with Neon Blue and Neon Red Joy‑Con - HAC-001(-01)', Decimal('299.00'), 'Sports & Outdoors', 407, 108884),
 ('B071WC2NK8', "Men's Full-Zip Eco-Smart Hoodie", Decimal('15.99'), 'Sports & Outdoors', 379, 87665),
 ('B006QF3TW4', 'LifeStraw Personal Water Filter for Hiking, Camping, Travel, and Emergency Preparedness', Decimal('17.47'), 'Toys & Games', 361, 90333),
 ('B07F6F631N', 'Reynolds Kitchens Parchment Paper Roll, 60 Square Feet', Decimal('3.77'), 'Appliances', 356, 41622),
 ('B0725GYNG6', 'Amazon Basics Silicone, Non-Stick, Food Safe Baking Mat - Pack of 2', Decimal('13.15'), 'Toys & Games', 333, 82711),
 ('B0911WZHKX', "Fruit of the Loom Women's Built Up Tank Style Sports Bra", Decimal('15.66'), 'Sports & Outdoors', 327, 80309),
 ('B07XLTNY8H', "Men's Regular Leg Boxer Briefs, Multipack", Decimal('17.99'), 'Pet Supplies', 324, 79988),
 ('B000G77PF0', 'Mens Loose Fit Heavyweight Short-sleeve Pocket T-shirt', Decimal('12.74'), 'Office Products', 287, 67459),
 ('B08BX7FV5L', 'Fire HD 10 tablet, 10.1", 1080p Full HD, 32 GB, latest model (2021 release), Black', Decimal('149.99'), 'Beauty & Personal Care', 265, 68465),
 ('B099YMHY1F', "Fruit of the Loom Women's Spaghetti Strap Cotton Pullover Sports Bra", Decimal('14.00'), 'Pet Supplies', 264, 61876),
 ('B01KWTGAVQ', 'Amazon Basics Reusable Silicone Baking Cups, Muffin Liners - Pack of 12, Multicolor', Decimal('7.39'), 'Clothing, Shoes & Jewelry', 248, 59812),
 ('B07GG1ZG8W', 'unisex-baby 8-pack Short Sleeve Onesies Bodysuits', Decimal('15.88'), 'Sports & Outdoors', 245, 58195),
 ('B099ZK84KS', "Women's Spaghetti Strap Cotton Pullover Sports Bra", Decimal('14.00'), 'Clothing, Shoes & Jewelry', 245, 61876),
 ('B018L2WM86', 'Vont 2 Pack LED Camping Lantern, Super Bright Portable Survival Lanterns, Must Have During Hurricane, Emergency, Storms, Outages, Original Collapsible Camping Lights/Lamp (Batteries Included)', Decimal('15.27'), 'Toys & Games', 237, 56768),
 ('B09JQMJHXY', 'Apple AirPods Pro', Decimal('174.00'), 'Pet Supplies', 220, 49154),
 ('B07815JCHQ', "Men's Tech 2.0 Short-Sleeve T-Shirt", Decimal('18.75'), 'Clothing, Shoes & Jewelry', 214, 61575),
 ('B01N5WNP8J', "Men's Multi-Pack Mesh Ventilating Comfort Fit Performance No-Show Socks", Decimal('25.00'), 'Pet Supplies', 212, 41349),
 ('B00AKSD004', 'Champion Men\'s 9" Jersey Short with Pockets', Decimal('11.63'), 'Toys & Games', 211, 48762)]

Analysis: The most popular product is the "Echo Dot 4th Gen" with almost 700 orders. It seems that in general, there is a correlation between review count and number of orders.

4. What % of products are either prime, sponsored, or Amazon's choice?

In [38]:
%%sql

SELECT
	SUM(is_prime)/COUNT(is_prime) AS percent_prime,
	SUM(is_sponsored)/COUNT(is_sponsored) AS percent_sponsored,
	SUM(is_amazon_choice)/COUNT(is_amazon_choice) AS percent_amazon_choice
FROM product;

 * mysql://admin:***@lmu-dev-sql.c7huj6y1fkyn.us-east-2.rds.amazonaws.com/sql_portfolio_project
1 rows affected.


[(Decimal('0.8785'), Decimal('0.1354'), Decimal('0.0052'))]

Analysis: Almost 90% of products on Amazon are Prime, and only 0.5% are "Amazon Choice." It would be interesting to see how this affects sales, review counts, and ratings.

5. Who are the top spending customers? Where are they from?

In [40]:
%%sql
SELECT
	c.id,
	SUM(co.order_total) AS total_spent,
	c.first_name,
	c.last_name,
    c.city,
    c.state
FROM customers c
JOIN customer_orders co
	ON c.id = co.cust_id
GROUP BY c.id
ORDER BY total_spent DESC
LIMIT 20;

 * mysql://admin:***@lmu-dev-sql.c7huj6y1fkyn.us-east-2.rds.amazonaws.com/sql_portfolio_project
20 rows affected.


[(243, Decimal('2479.20'), 'Gay', 'Epps', 'Inglewood', 'CA'),
 (348, Decimal('2417.22'), 'Barby', 'Rottcher', 'Philadelphia', 'PA'),
 (339, Decimal('2417.22'), 'Nicole', 'Klimentyev', 'El Paso', 'TX'),
 (770, Decimal('2386.23'), 'Simone', 'Ticksall', 'Nashville', 'TN'),
 (271, Decimal('2293.26'), 'Tawnya', 'Picford', 'Lubbock', 'TX'),
 (461, Decimal('2293.26'), 'Pen', 'Heaviside', 'Louisville', 'KY'),
 (961, Decimal('2262.27'), 'Armando', 'Pimbley', 'Augusta', 'GA'),
 (548, Decimal('2231.28'), 'Gui', 'Satyford', 'Lehigh Acres', 'FL'),
 (753, Decimal('2200.29'), 'Lizabeth', 'Brisley', 'Baltimore', 'MD'),
 (213, Decimal('2200.29'), 'Bryn', 'Yearsley', 'Cambridge', 'MA'),
 (128, Decimal('2200.29'), 'Baryram', 'Matevushev', 'Cincinnati', 'OH'),
 (403, Decimal('2169.30'), 'Alphard', 'Hare', 'San Antonio', 'TX'),
 (717, Decimal('2169.30'), 'Tessa', 'Stych', 'Wilmington', 'DE'),
 (875, Decimal('2138.31'), 'Heall', 'MacPaike', 'Houston', 'TX'),
 (65, Decimal('2138.31'), 'Tilly', 'Haylett', 'Savannah', 'GA'),
 (47, Decimal('2138.31'), 'Lew', 'Watman', 'Jamaica', 'NY'),
 (874, Decimal('2107.32'), 'Edwin', 'Lowndesbrough', 'Nashville', 'TN'),
 (250, Decimal('2107.32'), 'Ode', 'Jacomb', 'Richmond', 'VA'),
 (508, Decimal('2107.32'), 'Oriana', 'Attwell', 'Garden Grove', 'CA'),
 (119, Decimal('2076.33'), 'Mitzi', 'Honywill', 'Minneapolis', 'MN')]

Analysis: There does not seem to be any one customer that is an outlier in this case. The top buyers have spent similar amounts of money, however there is significant Texas representation on this list.

## Analysis

### Primary Question: What are the most popular products within each category? Are they Prime enabled, sponsored, or "Amazon Choice" products?

Justification: It is important to know the most popular product in each category as these are the leading brands within their domains, and account for a huge chunk of sales. Also, it is interesting to see how significant "Amazon Choice", "Prime", and sponsorship impact the success of products.

Features Used: View, Window Function, Join, Group By 

In [17]:
%%sql

CREATE VIEW product_rank AS
	SELECT
		p.category,
		RANK() OVER(
			PARTITION BY p.category
			ORDER BY SUM(co.order_total) DESC
		) AS ranking,
		p.asin,
		p.is_prime,
		p.is_sponsored,
		p.is_amazon_choice,
		p.title,
		SUM(co.order_total) AS total_spent,
		COUNT(co.asin) AS num_orders
	FROM product p
	JOIN customer_orders co 
		ON p.asin = co.asin
	GROUP BY category, asin;

 * mysql://admin:***@lmu-dev-sql.c7huj6y1fkyn.us-east-2.rds.amazonaws.com/sql_portfolio_project
0 rows affected.


[]

In [19]:
%%sql

SELECT *
FROM product_rank
WHERE ranking BETWEEN 1 AND 3;

 * mysql://admin:***@lmu-dev-sql.c7huj6y1fkyn.us-east-2.rds.amazonaws.com/sql_portfolio_project
27 rows affected.


[('Appliances', 1, 'B0168CNYRE', 1, 0, 0, 'Dash Rapid Egg Cooker: 6 Egg Capacity Electric Egg Cooker for Hard Boiled Eggs, Poached Eggs, Scrambled Eggs, or Omelets with Auto Shut Off Feature - Red', Decimal('28014.96'), 415),
 ('Appliances', 2, 'B07F6F631N', 1, 0, 0, 'Reynolds Kitchens Parchment Paper Roll, 60 Square Feet', Decimal('24110.22'), 356),
 ('Appliances', 3, 'B07TSCZV2P', 1, 0, 0, 'Lemedy Women Padded Sports Bra Fitness Workout Running Shirts Yoga Tank Top', Decimal('11931.15'), 177),
 ('Beauty & Personal Care', 1, 'B07XJ8C8F7', 1, 0, 0, 'Echo Dot (4th Gen) | Smart speaker with clock and Alexa | Glacier White', Decimal('44408.67'), 682),
 ('Beauty & Personal Care', 2, 'B08BX7FV5L', 1, 0, 0, 'Fire HD 10 tablet, 10.1", 1080p Full HD, 32 GB, latest model (2021 release), Black', Decimal('18903.90'), 265),
 ('Beauty & Personal Care', 3, 'B07MYTBV6N', 1, 0, 0, 'TEMI Dinosaur Toys for Kids 3-5 with Activity Play Mat & Trees, Educational Realistic Dinosaur Play Set to Create a Dino World Including T-Rex, Triceratops, Velociraptor, Great Gift for Boys & Girls', Decimal('9204.03'), 135),
 ('Clothing, Shoes & Jewelry', 1, 'B01KWTGAVQ', 1, 0, 0, 'Amazon Basics Reusable Silicone Baking Cups, Muffin Liners - Pack of 12, Multicolor', Decimal('16889.55'), 248),
 ('Clothing, Shoes & Jewelry', 2, 'B07815JCHQ', 1, 0, 0, "Men's Tech 2.0 Short-Sleeve T-Shirt", Decimal('15030.15'), 214),
 ('Clothing, Shoes & Jewelry', 3, 'B078GFYJKL', 1, 0, 0, "Men's Powerblend Fleece Pullover Hoodie, Script Logo", Decimal('14286.39'), 198),
 ('Electronics', 1, 'B0005ZXPY8', 0, 0, 0, 'Arm & Hammer Baking Soda-4LB (01170)', Decimal('11838.18'), 165),
 ('Electronics', 2, 'B01E3EHRFY', 1, 0, 0, 'Wise Owl Outfitters Camping Hammock - Portable Hammock Single or Double Hammock Camping Accessories for Outdoor, Indoor w/ Tree Straps', Decimal('11032.44'), 174),
 ('Electronics', 3, 'B099VMT8VZ', 1, 0, 0, 'Oculus Quest 2 — Advanced All-In-One Virtual Reality Headset — 128 GB', Decimal('8615.22'), 137),
 ('Home & Kitchen', 1, 'B00339910A', 1, 0, 0, 'Coleman Camping Chair with Built-in 4 Can Cooler', Decimal('12519.96'), 204),
 ('Home & Kitchen', 2, 'B01DR6ZN0K', 1, 0, 0, 'Wilton Easy Layers Sheet Cake Pan, 2-Piece Set, Rectangle Steel Sheet Pan', Decimal('8398.29'), 135),
 ('Home & Kitchen', 3, 'B00L5YL66M', 1, 0, 0, "Hanes Women's Get Cozy Pullover ComfortFlex Fit Wirefree Bra MHG196", Decimal('5888.10'), 98),
 ('Office Products', 1, 'B000G77PF0', 1, 0, 0, 'Mens Loose Fit Heavyweight Short-sleeve Pocket T-shirt', Decimal('18686.97'), 287),
 ('Office Products', 2, 'B016YKK3LA', 1, 0, 0, "Women's Jersey Short", Decimal('13573.62'), 187),
 ('Office Products', 3, 'B07T86S72Q', 0, 0, 0, 'Toy Story 4', Decimal('11032.44'), 170),
 ('Pet Supplies', 1, 'B07XLTNY8H', 1, 0, 0, "Men's Regular Leg Boxer Briefs, Multipack", Decimal('20174.49'), 324),
 ('Pet Supplies', 2, 'B099YMHY1F', 1, 0, 0, "Fruit of the Loom Women's Spaghetti Strap Cotton Pullover Sports Bra", Decimal('15990.84'), 264),
 ('Pet Supplies', 3, 'B09JQMJHXY', 1, 0, 0, 'Apple AirPods Pro', Decimal('14441.34'), 220),
 ('Sports & Outdoors', 1, 'B07VGRJDFY', 1, 0, 0, 'Nintendo Switch with Neon Blue and Neon Red Joy‑Con - HAC-001(-01)', Decimal('25814.67'), 407),
 ('Sports & Outdoors', 2, 'B071WC2NK8', 1, 0, 0, "Men's Full-Zip Eco-Smart Hoodie", Decimal('24234.18'), 379),
 ('Sports & Outdoors', 3, 'B0911WZHKX', 1, 0, 0, "Fruit of the Loom Women's Built Up Tank Style Sports Bra", Decimal('20794.29'), 327),
 ('Toys & Games', 1, 'B006QF3TW4', 1, 0, 0, 'LifeStraw Personal Water Filter for Hiking, Camping, Travel, and Emergency Preparedness', Decimal('22250.82'), 361),
 ('Toys & Games', 2, 'B0725GYNG6', 1, 0, 0, 'Amazon Basics Silicone, Non-Stick, Food Safe Baking Mat - Pack of 2', Decimal('21042.21'), 333),
 ('Toys & Games', 3, 'B018L2WM86', 1, 0, 0, 'Vont 2 Pack LED Camping Lantern, Super Bright Portable Survival Lanterns, Must Have During Hurricane, Emergency, Storms, Outages, Original Collapsible Camping Lights/Lamp (

The Echo Dot (4th Gen) seems to be the top selling product total. Interestingly enough, NONE of the top products in any category are sponsored or Amazon Choice.

Recommendation: Look more into the effects of sponsorship and Amazon Prime on the success of a product.

### Related Question #1: How do the ratings of Prime, Sponsored, and Amazon Choice products compare to the average of all products?

Justification: In order to make sure that our business model is providing vendors with value, we should show that these qualifications are something that vendors benefit from so that they continue to sell their products on Amazon.

Features used: Common Table Expression (CTE), Case, Subquery

In [20]:
%%sql
WITH ratings AS(
	SELECT
		p.asin,
		p.is_prime,
		p.is_sponsored,
		p.is_amazon_choice,
		p.rating,
		AVG(p.rating) OVER() AS avg_rating,
		p.title
	FROM product p
), rating_comparison AS (
SELECT
	*,
	CASE WHEN rating > avg_rating THEN 1 ELSE 0 END AS above_avg
	FROM ratings
)
SELECT
	avg_rating,
	(SELECT 
		AVG(rating)
	FROM rating_comparison
	WHERE is_prime = 1) AS prime_avg_rating,
	(SELECT
		CONCAT(ROUND(SUM(above_avg)/COUNT(above_avg) * 100, 2), '%')
	FROM rating_comparison 
	WHERE is_prime = 1) AS prime_products_above_avg,
	(SELECT
		AVG(rating)
	FROM rating_comparison
	WHERE is_sponsored = 1) AS sponsored_avg_rating,
	(SELECT
		CONCAT(ROUND(SUM(above_avg)/COUNT(above_avg) * 100, 2), '%')
	FROM rating_comparison
	WHERE is_sponsored = 1) AS sponsored_products_above_avg,
	(SELECT
		AVG(rating)
	FROM rating_comparison
	WHERE is_amazon_choice = 1) AS choice_avg_rating,
	(SELECT
		CONCAT(ROUND(SUM(above_avg)/COUNT(above_avg) * 100, 2), '%')
	FROM rating_comparison
	WHERE is_amazon_choice = 1) AS choice_products_above_avg
FROM rating_comparison
LIMIT 1;

 * mysql://admin:***@lmu-dev-sql.c7huj6y1fkyn.us-east-2.rds.amazonaws.com/sql_portfolio_project
1 rows affected.


[(Decimal('4.20677'), Decimal('4.27945'), '83.60%', Decimal('3.54744'), '65.38%', Decimal('4.50000'), '100.00%')]

This one took a while.

It seems that Prime products have a slightly higher average rating than all Amazon products, and 83.6% of them have a higher rating than the average. This indicates that simply by selling your product with Amazon Prime, you will increase customer satisfaction with the product.

Sponsored products have a much lower average rating than global average, probably because anyone with money can pay for an advertisement.

100% of "Amazon's Choice" products have above a 4.20677 rating! However, I later discovered that the dataset only has 3 such products... so it is not enough to make an accurate assessment. If this trend continued, is would be very impressive and show that products labeled as "Amazon's Choice" are generally high quality and trustworthy.

Recommendation: Run the analysis again on a bigger dataset to see if the trend holds and perhaps make sure that sponsored products work as intended before allowing the advertisement to run. Also, study the data to see how these factors impact the actual sales of products.

### Related Question #2: How does Prime, Sponsorship, and Amazon Choice affect the likelihood of purchase?

Justification: While Prime and Choice products are highly rated, does this translate to increased quantities sold? It is important to see how well these products sell so we can see if consumers trust Amazon to make recommendations to them.

Features used: Common Table Expression(CTE), Subquery, Join

In [5]:
%%sql

WITH special_orders AS (
	SELECT
		co.order_id,
		co.asin,
		p.is_prime,
		p.is_sponsored,
		p.is_amazon_choice
	FROM customer_orders co 
	JOIN product p 
		ON co.asin = p.asin
)
SELECT 
	(
		SELECT CONCAT(ROUND(SUM(is_prime)/COUNT(asin) * 100, 2), '%')
		FROM product p
	) AS pct_products_prime,
	CONCAT(ROUND(SUM(s.is_prime)/COUNT(s.order_id) * 100, 2), '%') AS pct_orders_prime,
	(
		SELECT CONCAT(ROUND(SUM(is_sponsored)/COUNT(asin) * 100, 2), '%')
		FROM product p
	) AS pct_products_sponsored,
	CONCAT(ROUND(SUM(s.is_sponsored)/COUNT(s.order_id) * 100, 2), '%') AS pct_orders_sponsored,
	(
		SELECT CONCAT(ROUND(SUM(is_amazon_choice)/COUNT(asin) * 100, 2), '%')
		FROM product p
	) AS pct_products_choice,
	CONCAT(ROUND(SUM(s.is_amazon_choice)/COUNT(s.order_id) * 100, 2), '%') AS pct_orders_choice
FROM special_orders s;


 * mysql://admin:***@lmu-dev-sql.c7huj6y1fkyn.us-east-2.rds.amazonaws.com/sql_portfolio_project
1 rows affected.


[('87.85%', '95.46%', '13.54%', '3.52%', '0.52%', '0.20%')]

People love Prime products, which account for over 95% of all orders despite representing under 90% of Amazon's total catalogue. However, people seem to dislike sponsored products, which account for less than 4% of orders even though sponsored products make up over 13% of all products in the dataset.

Most interestingly, even though they are reviewed higher, the number of orders of "Amazon Choice" products does not reflect the percentage of products that are "Amazon Choice." This could be because people don't trust Amazon to recommend them products, Amazon's recommendations do not span to the most popular category of product, or that Amazon is not recommending products that customers want. 

Recommendation: Identify why "Amazon's Choice" is not seeing the number of sales that it should. Find new products in popular product categories that may be underrepresented in "Amazon's Choice" products and add them to the list.